applications/vanilla/models/class.discussionmodel.php

```diff
@@ -2464,4 +2464,38 @@ public static function allowedSortFields($Allowed = null) {
 
         return self::$AllowedSortFields;
     }
+
+
+    /**
+     * Tests whether a user has permission to view a specific discussion by checking category-specific permissions.
+     * Fires an event that can override the category view permission.
+     *
+     * @param object|integer $discussion The discussion ID or the discussion to test.
+     * @param string|integer $userId The ID of the user to test permission for. If empty, it defaults to Session user.
+     * @param string $categoryPermission The category permission to test against the user.
+     * @return bool Whether the user can view the discussion.
+     * @throws Exception
+     */
+    public function canViewDiscussion($discussion, $userId = '', $categoryPermission = 'Vanilla.Discussions.View') {
+        // Default to session user.
+        if (!$userId) {
+            $userId = val('UserID', Gdn::session(), false);
+        }
+        // Fetch discussion.
+        if (is_numeric($discussion)) {
+            $discussion = $this->getID($discussion);
+        }
+        $userModel = new UserModel();
+        // Get category permission.
+        $canView = $userId && $userModel->getCategoryViewPermission($userId, val('CategoryID', $discussion), $categoryPermission);
+
+        // Fire event to override permission ruling.
+        $this->EventArguments['discussion'] = $discussion;
+        $this->EventArguments['userId'] = $userId;
+        $this->EventArguments['categoryPermission'] = $categoryPermission;
+        $this->EventArguments['canView'] = &$canView;
+        $this->fireEvent('viewPermission');
```

Comment:
I'd rather the event arguments stay uppercase to be consistent with the rest of the framework. This isn't something we can easily change.

Also, We use ID, not Id and that won't change.